igwn_auth_utils/requests.py

```diff
@@ -0,0 +1,289 @@
+
+    fail_if_noauth : `bool`, optional
+        raise an `Exception` if no authorisation credentials are presented
+        or discovered.
+
+    Raises
+    ------
+    ValueError
+        if ``fail_if_noauth=True`` is given and no authorisation
+        token/credentials are presented or discovered.
+
+    RuntimeError
+        if ``cert=True`` or ``token=True`` is given and the relevant
+        credential was not actually discovered.
+
+    Note
+    ----
+    This class requires `Requests <https://requests.readthedocs.io/>`__.
+
+    See also
+    --------
+    requests.Session
+        for details of the standard options
+    """.strip()
+
+
+class SessionAuthMixin:
+    """Mixin for :class:`requests.Session` to add support for IGWN auth.
+
+    By default this mixin will automatically attempt to discover/configure
+    a bearer token (scitoken) or an X.509 credential, with options to
+    require/disable either of those, or all authentication entirely.
+
+    {parameters}
+    """
+    def __init__(
+        self,
+        auth=None,
+        cert=None,
+        token=None,
+        token_audience=None,
+        token_scope=None,
+        url=None,
+        force_noauth=False,
+        fail_if_noauth=False,
+        **kwargs,
+    ):
+        # initialise session
+        super().__init__(**kwargs)
+
+        # handle options
+        if force_noauth and fail_if_noauth:
+            raise ValueError(
+                "cannot select both force_noauth and fail_if_noauth",
+            )
+        if force_noauth:
+            return
+
+        # find creds if we can
+        self._init_auth(
+            auth=auth,
+            cert=cert,
+            token=token,
+            token_audience=token_audience,
+            token_scope=token_scope,
+            url=url,
+        )
+
+        # if no auth was found, and we need it, fail here
+        if not self._has_igwn_auth and fail_if_noauth:
+            raise ValueError("no valid authorisation credentials found")
+
+    def _init_auth(
+        self,
+        auth=None,
+        cert=None,
+        token=None,
+        token_audience=None,
+        token_scope=None,
+        url=None,
+    ):
+        # -- bearer token
+        tokenhdr = self._find_token_header(
+            token,
+            token_audience,
+            token_scope,
+        )
+        if tokenhdr:
+            self.headers["Authorization"] = tokenhdr
+            return
+
+        # -- cert auth
+        self.cert = self._find_x509_credentials(cert)
+        if self.cert:
+            return
+
+        # -- basic auth (netrc)
+        self.auth = self._find_username_password(url)
+
+    @property
+    def _has_igwn_auth(self):
+        """`True` if this `Session` is configured for IGWN authorization,
+        otherwise `False`.
+
+        :type: `bool`
+        """
+        return any((
+            self.auth,
+            "Authorization" in self.headers,
+            self.cert,
+        ))
```

Comment:
```suggestion
        return bool(
            self.auth or
            "Authorization" in self.headers or
            self.cert
        )
```

This is a micro-optimization to take advantage of short-circuit evaluation.

Reply:
Done in latest commit.